api/src/main/java/org/openmrs/liquibase/ChangeLogDetective.java

```diff
@@ -0,0 +1,161 @@
+			
+			for (String filename : changeSets) {
+				Liquibase liquibase = liquibaseProvider.getLiquibase(filename);
+				List<ChangeSet> rawUnrunChangeSets = liquibase.listUnrunChangeSets(new Contexts(context),
+				    new LabelExpression());
+				List<ChangeSet> refinedUnrunChangeSets = excludeVintageChangeSets(rawUnrunChangeSets);
+				
+				log.info("file '{}' contains {} un-run change sets", filename, refinedUnrunChangeSets.size());
+				logUnrunChangeSetDetails(filename, refinedUnrunChangeSets);
+				
+				unrunChangeSetsCount += refinedUnrunChangeSets.size();
+			}
+			
+			if (unrunChangeSetsCount == 0) {
+				log.info(
+				    "the Liquibase snapshot version that had been used to initialize the OpenMRS database is '{}'",
+				    version
+				);
+				return version;
+			}
+		}
+		
+		throw new IllegalStateException(
+		        "identifying the snapshot version that had been used to initialize the OpenMRS database failed as no candidate change set resulted in zero un-run changes");
+	}
+	
+	/**
+	 * Returns a list of Liquibase update files that contain un-run change sets.
+	 *
+	 * @param snapshotVersion the snapshot version that had been used to initialise the OpenMRS database
+	 * @param liquibaseProvider provides access to a Liquibase instance
+	 * @return a list of Liquibase update files that contain un-run change sets.
+	 * @throws Exception
+	 */
+	public List<String> getUnrunLiquibaseUpdateFileNames(String snapshotVersion, String context,
+	        LiquibaseProvider liquibaseProvider) throws Exception {
+		List<String> unrunLiquibaseUpdates = new ArrayList<>();
+		
+		List<String> updateVersions = changeLogVersionFinder.getUpdateVersionsGreaterThan(snapshotVersion);
+		List<String> updateFileNames = changeLogVersionFinder.getUpdateFileNames(updateVersions);
+		
+		for (String filename : updateFileNames) {
+			Liquibase liquibase = liquibaseProvider.getLiquibase(filename);
+			List<ChangeSet> unrunChangeSets = liquibase.listUnrunChangeSets(new Contexts(context), new LabelExpression());
+			log.info("file '{}}' contains {} un-run change sets", filename, unrunChangeSets.size());
+			
+			if (unrunChangeSets.size() > 0) {
+				unrunLiquibaseUpdates.add(filename);
+			}
+		}
+		
+		return unrunLiquibaseUpdates;
+	}
+	
+	List<String> getSnapshotVersionsInDescendingOrder(Map<String, List<String>> snapshotCombinations) {
+		List<String> versions = new ArrayList<>();
+		versions.addAll(snapshotCombinations.keySet());
+		Collections.sort(versions, Collections.reverseOrder());
+		return versions;
+	}
+	
+	List<ChangeSet> excludeVintageChangeSets(List<ChangeSet> changeSets) {
+		List<ChangeSet> result = new ArrayList<>();
+		for (ChangeSet changeSet : changeSets) {
+			if (!isVintageChangeSet(changeSet)) {
+				result.add(changeSet);
+			}
+		}
+		return result;
+	}
+	
+	boolean isVintageChangeSet(ChangeSet changeSet) {
+		if (changeSet.getId().equals(DISABLE_FOREIGN_KEY_CHECKS) && changeSet.getAuthor().equals(BEN)) {
```

Comment:
What happens if we get new changesets with the same author name?

Reply:
That is a valid point, I can narrow that check down to change sets in `liquibase-core-data-1.9.x.xml`, that shall make it future proof.

Reply:
Done.